### DISCUSSION TITLE
rtfunicode does not support flag emoji, causes crash, so what is better way to avoid?

### DISCUSSION THREAD

**Discussion by Kurdakov** [ThioJoe/YT-Spammer-Purge#257] [General]
somehow author of rtfunicode which is used in YT-Spammer-Purge did not realise support for multi unicode character emoji, so comments like this  https://www.youtube.com/post/Ugz9kkSrsvSqZaW6Kjd4AaABCQ?lc=UgwFHf-8X76qm2DUh494AaABAg (see flag - it is shown differently in browsers on Windows and Linux/Max - on Linux/Mac correctly, on Windows just shows first flag character and, btw this comment  is false positive, but was offered to save in log), but the end result is the same.
rtfunicode.py line 37 

 _charescape.sub(_replace, text).encode('ascii', errors)

will cause crash.
UnicodeEncodeError: 'ascii' codec can't encode characters in position 123-128: ordinal not in range(128)

seems also black woman emoji (two unicode characters) will cause the same crash.

so how to proceed?
if it is ok to save logs without emojis, then currently it's possible just filter flag emojis
maybe like some part from here: https://titanwolf.org/Network/Articles/Article?AID=77e37ced-15fe-49a1-b0c8-5201ede65cd6
if not, possibly there is need in another format to save rich logs (possibly 'html log' - basically header with list  of comments)
maybe visitors have some good suggestions?

**Comment by rbeer** [comment]
Problems like this are why logs should generally be raw text. Colorization, interpretation of emojis, etc. should be left to the processing program.

**Reply by TechStudent10** [reply]
Yeah but I think ThioJoe did this because...

IT HAS COLORS.

But no I'm being serious, they could be really helpful when navigating. Still don't know how to fix this issue tho...

**Reply by rbeer** [reply]
> Still don't know how to fix this issue tho...

The author of rtfunicode has written a detailed blog post about why and how he created the module: https://www.zopatista.com/python/2012/06/06/rtf-and-unicode/

> This control sequence can properly represent unicode, at least for the U+0000 through to U+FFFF codepoints.

So the first question is, do emojis work in RTF at all? They are all beyond U+FFFF.

The next thing is that different RTF readers use different declaration styles. This is th US flag emoji in an .rtf document, saved by LIbreOffice writer: `\uc0 \u55356\uc1 \u56826\'3f\uc0 \u55356\uc1 \u56824\'3f}`. Entirely different from the style that the module produces. Probably because the RTF spec only covers the codepages up to U+FFFF, so they had to come up with their own non-standard way to support everything beyond. This is me just guessing, though. Also, not every RTF reader even displays those emojis; writer doesn't:
![us-flag](https://user-images.githubusercontent.com/3411649/147724819-1e4370cf-fb91-40aa-848b-0a0c4d039015.png)


No offense, but this is the programmer's version of the "it's the current year" meme. Don't use formats that don't support UTF, unless some legacy system forces you to. You just buy into heaps of headaches like this.
If you really want to save the logs in a rich text format, I'd go with @Kurdakov 's suggestion and use HTML.
